hazelcast/src/main/java/com/hazelcast/cache/impl/PreJoinCacheConfig.java

```diff
@@ -56,7 +57,12 @@ public PreJoinCacheConfig(CacheConfig cacheConfig) {
     }
 
     public PreJoinCacheConfig(CacheConfig cacheConfig, boolean resolved) {
-        cacheConfig.copy(this, resolved);
+        this(cacheConfig, resolved, null);
+    }
+
+    public PreJoinCacheConfig(CacheConfig cacheConfig, boolean resolved,
+            SerializationService backupSerializationService) {
```

Comment:
I think we can remove this whole `backupSerializationService`. The case which you seem to be referring to is when the `CacheConfig` is constructed using `public CacheConfig(CacheSimpleConfig simpleConfig)`. I believe that if we set the serialization service in that constructor, it will always be non-null. Do you have a test (either on our side or on your side) which tests for this NPE?

Reply:
No, this was during reSerialize() which does a copy() of the cache config, not simpleConfig

Reply:
I am sure it can be removed provided serializationService is made sure never to be null

Reply:
Yes, `reSerialize()` was doing a copy of a `CacheConfig` which was most probably constructed using `public CacheConfig(CacheSimpleConfig simpleConfig)` as that constructor does not ensure the serialization service is set.

Do you have a test (either on our side or on your side) which tests for this NPE?

Reply:
Maybe... this was so long ago I don't remember

Reply:
I'll remove it and we'll test then.

Reply:
> Yes, reSerialize() was doing a copy of a CacheConfig which was most probably constructed using public CacheConfig(CacheSimpleConfig simpleConfig)

I remember now. This is correct. If CacheConfig always have serializer non-null, the backup serialization code can be removed.

Reply:
I started removing it which then turned out to either the serialization service needing to be added to `public CacheConfig(CacheSimpleConfig simpleConfig)` or a setter added to `CacheConfig`. @vbekiaris is that constructor public API? 

Also, is `CacheConfig#copy` public API? Because we're changing the arguments to include `SerializationService`, which is private API.

Reply:
Pragmatically, I don't expect any end users actually use `CacheConfig(CacheSimpleConfig simpleConfig)` or `CacheConfig#copy`, however both are exposed as public. A workaround could be to add a package-private setter for `SerializationService` and access it via `CacheConfigAccessor` which is already declared as `@PrivateApi`.

Reply:
Changed it to your suggestion and removed backup serialization service. CacheConfig should now always have the serialization service set.